runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -348,16 +348,33 @@ public void prefixTrim(final Token address) {
                                     .flatMap(stripe -> stripe.getLogServers().stream())
                                     .map(e::getLogUnitClient)
                                     .map(client -> client.prefixTrim(address))
-                                    .forEach(CFUtils::getUninterruptibly);
-                            return null;    // No return value
-                        }, true);
+                                    .forEach(cf -> {CFUtils.getUninterruptibly(cf,
+                                            NetworkException.class, TimeoutException.class,
+                                            WrongEpochException.class);
+                                    });
+                            return null;
+                }, true);
                 // TODO(Maithem): trimCache should be epoch aware?
                 runtime.getSequencerView().trimCache(address.getSequence());
                 break;
             } catch (NetworkException | TimeoutException e) {
                 log.warn("prefixTrim: encountered a network error on try {}", x, e);
                 Duration retryRate = runtime.getParameters().getConnectionRetryRate();
                 Sleep.sleepUninterruptibly(retryRate);
+            } catch (WrongEpochException wee) {
+                log.warn("prefixTrim[{}]: wrongEpochException", address);
+                long runtimeEpoch = runtime.getLayoutView().getLayout().getEpoch();
+                long serverEpoch = wee.getCorrectEpoch();
+                // Only retry if wrongEpochException corresponds to message epoch
+                if (runtimeEpoch != serverEpoch) {
+                    log.warn("prefixTrim[{}]: wrongEpochException, runtime is in epoch {}, " +
+                            "while server is in epoch {}. Invalidate layout for this client " +
+                            "and retry, attempt: {}/{}", address, runtimeEpoch, serverEpoch, x+1, numRetries);
+                    runtime.invalidateLayout();
+                } else {
+                    // wrongEpochException corresponds to a stale trim address (prefix trim token on the wrong epoch)
```

Comment:
If it's a stale trim address, we should get a `StaleTokenException`, right?  If our current code is actually throwing a `WrongEpochException` in this case, I think it's not appropriate and should be fixed. what do you think?

Reply:
Stale token and wrong epoch are not the same thing. The prior is specific to transaction tokens.

Reply:
@Maithem I saw the only place to throw `StaleTokenException` is in `AddressSpaceView::write()`, so I think it's not transaction specific, right?